contracts/exchange/test/internal.ts

```diff
@@ -48,16 +50,358 @@ blockchainTests('Exchange core internal functions', env => {
+        orderFactoryRight = new OrderFactory(privateKeyRight, defaultOrderParamsRight);
+    });
+
+    blockchainTests('assertValidMatch', () => {
+        it('should revert if the prices of the left order is less than the price of the right order', async () => {
+            const leftOrder = await orderFactoryLeft.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(49, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+            });
+            const rightOrder = await orderFactoryRight.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+            });
+            const orderHashHexLeft = orderHashUtils.getOrderHashHex(leftOrder);
+            const orderHashHexRight = orderHashUtils.getOrderHashHex(rightOrder);
+            const expectedError = new ExchangeRevertErrors.NegativeSpreadError(orderHashHexLeft, orderHashHexRight);
+            return expect(testExchange.assertValidMatch.callAsync(leftOrder, rightOrder)).to.revertWith(expectedError);
+        });
+
+        it('should succeed if the prices of the left and right orders are equal', async () => {
+            const leftOrder = await orderFactoryLeft.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+            });
+            const rightOrder = await orderFactoryRight.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+            });
+            return expect(testExchange.assertValidMatch.callAsync(leftOrder, rightOrder)).to.be.fulfilled('');
+        });
+
+        it('should succeed if the price of the left order is higher than the price of the right', async () => {
+            const leftOrder = await orderFactoryLeft.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+            });
+            const rightOrder = await orderFactoryRight.newSignedOrderAsync({
+                makerAssetAmount: Web3Wrapper.toBaseUnitAmount(100, 18),
+                takerAssetAmount: Web3Wrapper.toBaseUnitAmount(50, 18),
+            });
+            return expect(testExchange.assertValidMatch.callAsync(leftOrder, rightOrder)).to.be.fulfilled('');
+        });
+    });
+
+    blockchainTests('calculateCompleteFillBoth', () => {
+        /**
+         * Asserts that the actual result of calling `calculateCompleteFillBoth()` is as expected.
+         */
+        async function assertCalculateCompleteFillBothAsync(
+            expectedMatchedFillResults: MatchedFillResults,
```

Comment:
I don't think this pattern of passing in explicit results really works for us, because it is so unmaintainable. I would rather we implement *well-tested* reference functions (see `../src/reference_functions.ts`). This would cut down the line count of this file by A LOT, and we'd only have to update the reference function when the implementation changes. Not to mention that reference functions will be useful when we upgrade some of the older integration tests.

Reply:
I'm not sure I agree with you on this one. In my opinion, well testing the reference functions would actually mean testing them the same as I test the real functions here. Otherwise, I don't think that the assertions we make are very strong, and it's easy to introduce logical errors into the reference functions. We should discuss this during standup.